### PR TITLE
Apim 1198 api v4 for members post

### PR DESCRIPTION
**Pull request by loriepisicchio**
## Issue

https://gravitee.atlassian.net/browse/APIM-1198

## Description

Implement POST + PUT for API members

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-rvyvaigmpq.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml [hunk 1]
```diff
@@ -258,6 +258,54 @@ paths:
                     $ref: "#/components/responses/MembersResponse"
                 default:
                     $ref: "#/components/responses/Error"
+        post:
+            tags:
+                - APIs
+                - API Members
+            summary: Add a member to an API or edit membership
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml]:
```suggestion
            summary: Add a member to an API
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml [hunk 2]
```diff
@@ -258,6 +258,54 @@ paths:
...
+            description: |
+                Add a new member to an API, or edit existing membership.
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml]:
```suggestion
                Add a new member to an API.
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml [hunk 3]
```diff
@@ -1438,6 +1486,31 @@ components:
...
+                roleName:
+                    type: string
+                    description: The name of the role
+                    minLength: 1
+        UpdateApiMembership:
+            type: object
+            properties:
+                memberId:
+                    type: string
+                    description: User's technical identifier
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml]:
```suggestion
                    description: Technical identifier for existing user.
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource.java [hunk 4]
```diff
@@ -61,4 +68,45 @@ public MembersResponse getApiMembers(@BeanParam PaginationParam paginationParam)
...
     }
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Permissions({ @Permission(value = RolePermission.API_MEMBER, acls = RolePermissionAction.CREATE) })
+    public Response createApiMembership(CreateApiMembership apiMembership) {
+        if (PRIMARY_OWNER.name().equals(apiMembership.getRoleName())) {
+            throw new SinglePrimaryOwnerException(RoleScope.API);
+        }
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource.java]:
```suggestion
        checkRoleIsNotPrimaryOwner(apiMembership.getRoleName());
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.Response;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.*;
+
+public class ApiMembersResource_CreateOrUpdateTest extends AbstractResourceTest {
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java]: question: should we split into `...createTest` and `...updateTest`?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    public void init() {
+        reset(membershipService);
+        GraviteeContext.cleanContext();
+        GraviteeContext.setCurrentEnvironment(ENVIRONMENT);
+        GraviteeContext.setCurrentOrganization(ORGANIZATION);
+
+        EnvironmentEntity environmentEntity = new EnvironmentEntity();
+        environmentEntity.setId(ENVIRONMENT);
+        environmentEntity.setOrganizationId(ORGANIZATION);
+        doReturn(environmentEntity).when(environmentService).findByOrgAndIdOrHrid(ORGANIZATION, ENVIRONMENT);
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java]: reminder for a future PR, maybe we could move this part into an `AbstractEnvironmentTest `?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        memberEntity.setDisplayName("John Doe");
+        var ownerRole = new RoleEntity();
+        ownerRole.setName("OWNER");
+        memberEntity.setRoles(List.of(ownerRole));
+
+        // Feel like I'm writing twice the same code : in the resource, and in the test.
+        // Do you have a better idea ?
+        MembershipService.MembershipReference reference = new MembershipService.MembershipReference(MembershipReferenceType.API, apiId);
+        MembershipService.MembershipMember member = new MembershipService.MembershipMember("userId", null, MembershipMemberType.USER);
+        MembershipService.MembershipRole role = new MembershipService.MembershipRole(RoleScope.API, "OWNER");
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java]: you don't use these variables, do you?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    public void should_edit_a_membership() {
+        var memberEntity = new MemberEntity();
+        memberEntity.setDisplayName("John Doe");
+        var ownerRole = new RoleEntity();
+        ownerRole.setName("OWNER");
+        memberEntity.setRoles(List.of(ownerRole));
+
+        // Feel like I'm writing twice the same code : in the resource, and in the test.
+        // Do you have a better idea ?
+        MembershipService.MembershipRole role = new MembershipService.MembershipRole(RoleScope.API, "OWNER");
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/test/java/io/gravitee/rest/api/management/v4/rest/resource/api/ApiMembersResource_CreateOrUpdateTest.java]: Do you use this variable ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml [hunk 9]
```diff
@@ -1438,6 +1486,31 @@ components:
...
+                externalReference:
+                    type: string
+                    description: User's reference for user provided from an identity provider.
+                roleName:
+                    type: string
+                    description: The name of the role
+                    minLength: 1
+        UpdateApiMembership:
+            type: object
+            properties:
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml:1825]: You should add `memberId` and `roleName` into a required field

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml [hunk 10]
```diff
@@ -1438,6 +1486,31 @@ components:
                     type: string
                     description: Role's name.
                     example: OWNER
+        CreateApiMembership:
+            type: object
+            description: Membership can be created for existing users, or for users provided by external identity provider. Thus, either userId or externalReference are required.
+            properties:
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v4/gravitee-apim-rest-api-management-v4-rest/src/main/resources/management-openapi-v4.yaml:1811]: You should add `roleName` into a required field

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java [hunk 11]
```diff
@@ -1726,6 +1726,55 @@ public List<MemberEntity> updateRolesToMemberOnReferenceBySource(
...
 
+    @Override
+    public MemberEntity createNewMembershipForApi(
+        ExecutionContext executionContext,
+        String apiId,
+        String userId,
+        String externalReference,
+        String roleName
+    ) {
+        MembershipService.MembershipReference reference = new MembershipService.MembershipReference(MembershipReferenceType.API, apiId);
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java:1737]: I think we should check that we have at least a userId or an externalReference, and send a 400 if it's not the case.
We could have both, but in that case, the userId would take over.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java [hunk 12]
```diff
@@ -1726,6 +1726,55 @@ public List<MemberEntity> updateRolesToMemberOnReferenceBySource(
...
+        if (member.getMemberId() != null) {
+            MemberEntity userMember = getUserMember(
+                GraviteeContext.getExecutionContext(),
+                MembershipReferenceType.API,
+                apiId,
+                member.getMemberId()
+            );
+            if (userMember != null && userMember.getRoles() != null && !userMember.getRoles().isEmpty()) {
+                membership = updateRoleToMemberOnReference(GraviteeContext.getExecutionContext(), reference, member, role);
+            }
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java]: As we are in a creation process, if the user is already associated to the API, we should maybe return a 400.
wdyt ?

**Review by phiz71 — CHANGES_REQUESTED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java (hunk 11 above)
Comment (loriepisicchio) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MembershipServiceImpl.java:1737]: Done at ressource level

**Review by phiz71 — APPROVED** [review]

**Pull request merged**